src/peft/tuners/vera/model.py

```diff
@@ -0,0 +1,746 @@
+
+
+class VeraModel(BaseTuner):
+    """
+    Creates Vector-based Random Matrix Adaptation (Vera) model from a pretrained transformers model.
+
+    Args:
+        model ([`~transformers.PreTrainedModel`]): The model to be adapted.
+        config ([`VeraConfig`]): The configuration of the Vera model.
+        adapter_name (`str`): The name of the adapter, defaults to `"default"`.
+
+    Returns:
+        `torch.nn.Module`: The Vera model.
+
+    Example:
+
+        ```py
+        >>> from transformers import AutoModelForSeq2SeqLM
+        >>> from peft import VeraModel, VeraConfig
+
+        >>> config = VeraConfig(
+        ...     task_type="SEQ_2_SEQ_LM",
+        ...     r=8,
+        ...     target_modules=["q", "v"],
+        ...     vera_dropout=0.01,
+        ... )
+
+        >>> model = AutoModelForSeq2SeqLM.from_pretrained("t5-base")
+        >>> vera_model = VeraModel(model, config, "default")
+        ```
+
+        ```py
+        >>> import transformers
+        >>> from peft import VeraConfig, PeftModel, get_peft_model
+
+        >>> target_modules = ["q_proj", "k_proj", "v_proj", "out_proj", "fc_in", "fc_out", "wte"]
+        >>> config = VeraConfig(
+        ...     r=4, target_modules=target_modules, vera_dropout=0.1, bias="none", task_type="CAUSAL_LM"
+        ... )
+
+        >>> model = transformers.GPTJForCausalLM.from_pretrained(
+        ...     "kakaobrain/kogpt",
+        ...     revision="KoGPT6B-ryan1.5b-float16",  # or float32 version: revision=KoGPT6B-ryan1.5b
+        ...     pad_token_id=tokenizer.eos_token_id,
+        ...     use_cache=False,
+        ...     device_map={"": rank},
+        ...     torch_dtype=torch.float16,
+        ...     load_in_8bit=True,
+        ... )
+        >>> model = prepare_model_for_int8_training(model)
+        >>> vera_model = get_peft_model(model, config)
+        ```
+
+    **Attributes**:
+        - **model** ([`~transformers.PreTrainedModel`]) -- The model to be adapted.
+        - **peft_config** ([`VeraConfig`]): The configuration of the Vera model.
+    """
+
+    def _find_first_dim(self) -> int:
+        for module in self.base_model.modules():
+            # TODO: check this is correct also for conv and embedding as base weight shape may be different
+            # TODO: could expand to return "first dim of each layer type"
+            if isinstance(module, VeraLayer):
+                return module.weight.shape[-1]
+
+    def __init__(self, model, config, adapter_name) -> None:
+        super().__init__(model, config, adapter_name)
+        config = config[adapter_name]
+        generator = torch.Generator(device="cpu").manual_seed(config.projection_prng_key)
+
+        first_vera_dim = self._find_first_dim()
+        vera_A = _kaiming_init((config.r, first_vera_dim), generator=generator)
+        vera_B = _kaiming_init((first_vera_dim, config.r), generator=generator)
+
+        self.register_buffer("vera_A", vera_A, persistent=config.save_projection)
```

Comment:
Good point about the individual adapter layers having access to all attributes that they need, independent of the `VeraModel`. This would also simplify the implementation.

The issue with this that we discussed earlier is that we would get one copy of the identical weight for each layer when we create the `state_dict`, right? I.e. we would need to clean up the `state_dict` when saving and loading. Or is there a way to avoid that?

Reply:
@pacman100 any thoughts on the above? If this is the case, it would be a significant disadvantage to your proposed change. Otherwise, it makes sense.